cmd/zdb/zdb.c

```diff
@@ -4187,6 +4196,264 @@ same_metaslab(spa_t *spa, uint64_t vdev, uint64_t off1, uint64_t off2)
+ *
+ * All three types of blocks are listed on a single line
+ *
+ * By default the table is printed in nicenumber format (e.g. 123K) but
+ * if the '-P' parameter is specified then the full raw number is printed
+ * out.
+ */
+static void
+dump_size_histograms(zdb_cb_t *zcb)
+{
+	/*
+	 * Keep track of the smallest and largest blocksize used in
+	 * this file system so that we don't print out the block
+	 * sizes that have zero count of blocks used.
+	 */
+	int minidx = SPA_MAXBLOCKSHIFT;
+	int maxidx = 0;
+	uint64_t max = 0;
+
+	/*
+	 * A temporary buffer that allows us to convert a number into
+	 * a string using zdb_nicenumber to allow either raw or human
+	 * readable numbers to be output.
+	 */
+	char numbuf[32];
+	int maxlen = 0;
+	int colwidth = 0;
+	int firstcolwidth = 0;
+
+	/* Make sure the output won't get truncated */
+	CTASSERT(sizeof (numbuf) >= NN_NUMBUF_SZ);
+
+	/*
+	 * Define titles which are used in the headers of the tables
+	 * printed by this routine.
+	 */
+	const char blocksize_title1[] = "block";
+	const char blocksize_title2[] = "size";
+	const char count_title[] = "Count";
+	const char length_title[] = "Length";
+	const char cumulative_title[] = "Cum.";
+
+	/*
+	 * Setup the histogram arrays (psize, lsize, and asize).
+	 */
+	one_histo_t parm_histo[NUM_HISTO];
+
+	parm_histo[0].name = "psize";
+	parm_histo[0].count = zcb->zcb_psize_count;
+	parm_histo[0].len = zcb->zcb_psize_len;
+	parm_histo[0].cumulative = 0;
+
+	parm_histo[1].name = "lsize";
+	parm_histo[1].count = zcb->zcb_lsize_count;
+	parm_histo[1].len = zcb->zcb_lsize_len;
+	parm_histo[1].cumulative = 0;
+
+	parm_histo[2].name = "asize";
+	parm_histo[2].count = zcb->zcb_asize_count;
+	parm_histo[2].len = zcb->zcb_asize_len;
+	parm_histo[2].cumulative = 0;
+
+	/*
+	 * Find the minimum and maximum rows to avoid printing
+	 * all zero rows at the beginning or end of the list.
+	 *
+	 * We are going to special case the instance of blocks
+	 * that are of ZERO bytes in length and report that only
+	 * as an anomaly/curiosity until we figure out why these
+	 * are occuring.
+	 *
+	 * At the same time compute the cumulative for each
+	 * histogram so we can fing the largest and insure there
+	 * are enough columns for printing that.
+	 */
+	for (int i = 0; i < SPA_MAXBLOCKSHIFT; i++) {
+		for (int j = 0;  j < NUM_HISTO; j++) {
+
+			parm_histo[j].cumulative += parm_histo[j].len[i];
+
+			if (parm_histo[j].len[i] > max)
+				max = parm_histo[j].len[i];
+
+			if (parm_histo[j].count[i] > 0 && i > maxidx)
+				maxidx = i;
+			if (parm_histo[j].count[i] > 0 && i < minidx)
+				minidx = i;
+
+			if (parm_histo[j].len[i] > 0 && i > maxidx)
+				maxidx = i;
+			if (parm_histo[j].len[i] > 0 && i < minidx)
+				minidx = i;
+		}
+	}
+	/*
+	 * The maximum number will be the largest of the cumulatives that
+	 * were computed just above.
+	 */
+	for (int j = 0; j < NUM_HISTO; j++) {
+		max = (parm_histo[j].cumulative > max) ?
+		    parm_histo[j].cumulative : max;
```

Comment:
nit: there's a `MAX(x, y)` which can be used for this, `max = MAX(parm_histo[j].cumulative, max);`.  Though I'd suggest a different variable name than `max` to avoid confusion.